### PR TITLE
Don't deep-copy read-only buffers on unpickle

### PR DESCRIPTION
**Pull request by crusaderky**
Fix use case where, when a task output which
- features _read-only_ pickle5 buffers, and
- is either not registered with `@dask_serialize.register` or is wrapped in a custom object

is transferred from a worker to another, its buffer is unnecessarily deep-copied.

Example:
```python
import distributed
import numpy as np
from collections import UserList

def f():
    a = np.arange(1024)
    a.flags.writeable = False
    return UserList([a])

client = distributed.Client(n_workers=2)
a, b = client.has_what()
x = client.submit(f, key="x", workers=[a])
y = client.submit(lambda _: None, x, key="y", workers=[b])
```

### CODE REVIEW

File: distributed/protocol/serialize.py [hunk 1]
```diff
@@ -86,19 +87,29 @@ def buffer_callback(f):
...
+
+    def ensure_writeable_flag(mv: memoryview, w: bool) -> memoryview:
+        if w and mv.readonly:
+            # Can't avoid a deep copy
+            # Note: this doesn't actually happen in real life, as both the comms
+            # layer and the SpillBuffer store everything in writeable bytearrays
+            return memoryview(bytearray(mv))
+        elif not w and not mv.readonly:
+            # Zero copy - this is just a flag
+            return mv.toreadonly()
```
Comment (crusaderky) [inline distributed/protocol/serialize.py:103]: This is the only functional change (replacing `memoryview(bytes(mv))`); the rest is cosmetic

File: distributed/protocol/serialize.py [hunk 2]
```diff
@@ -86,19 +87,27 @@ def buffer_callback(f):
...
+) -> Any:
+    pik, buffers = frames[0], frames[1:]
 
-    writeable = header.get("writeable")
-    if not writeable:
-        writeable = len(buffers) * (None,)
+    def ensure_writeable_flag(mv: memoryview, w: bool) -> memoryview:
+        if w and mv.readonly:
+            # Can't avoid a deep copy
+            return memoryview(bytearray(mv))
```
Comment (crusaderky) [inline distributed/protocol/serialize.py:98]: Adding an `assert False` to this code path makes a lot of :hankey: :hankey: :hankey: crop up to the surface.
Namely, https://github.com/dask/distributed/issues/8608

File: distributed/protocol/serialize.py [hunk 3]
```diff
@@ -86,19 +87,27 @@ def buffer_callback(f):
...
+            # Zero copy - this is just a flag
+            return mv.toreadonly()
+        else:
+            return mv
 
     buffers = [
-        memoryview(bytearray(mv) if w else bytes(mv)) if w == mv.readonly else mv
-        for w, mv in zip(writeable, map(ensure_memoryview, buffers))
+        ensure_writeable_flag(ensure_memoryview(mv), w)
+        for mv, w in zip(buffers, header["writeable"])
```
Comment (milesgranger) [inline distributed/protocol/serialize.py:107]: That's an easier read. :+1:

**Review by milesgranger — APPROVED** [review]
Looks good from my perspective, thanks @crusaderky.

**Pull request merged**